### DISCUSSION TITLE
What is the correct way to run model in a jax pmap?

### DISCUSSION THREAD

**Discussion by shahbuland** [google/flax#1431] [Q&A]
I'm quite new to distributed programming and jax. I'm trying to do gradient accumulation over some minibatches with a flax.linen model. I want the minibatches to be distributed across different TPU cores with pmap. I have my function that takes a single minibatch that I want to pmap, and I have my model params (which don't change until after all gradients have been calculated and accumulated). What is the correct way to use the model in the pmapped function? Two ideas I had so far (which run) were:  
1. Just use params from outside the functions scope, which seems wrong because then the function isn't "pure" like jax docs specify.  
2. Set param tree as static through static_broadcasted_argnum in pmap, which seems wrong because the docs say that is for compile time constants, which the model params aren't.

**Comment by andsteing** [comment]
You would first replicate the parameters with `flax.jax_utils.replicate()` and then have them as a normal parameter to your `pmap()`ped function.

Example code:

https://github.com/google/flax/blob/f8dc89a6324506455e5d82e5b90fe2d8ee4edac9/examples/imagenet/train.py#L325-L339

Note that the return values from a `pmap()`ped function are also "replicated" and if you want to do something with them in the main program (e.g. storing a checkpoint) you would first remove that extra replication by `flax.jax_utils.unreplicate()`.
